user_tools/src/spark_rapids_pytools/resources/dev/prepackage_mgr.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module defines the logic to fetch all dependencies and resources required to run the user-tools
+without need to access the web during runtime.
+"""
+
+import os
+import shutil
+import tarfile
+from concurrent.futures import ThreadPoolExecutor
+from typing import Optional
+
+import fire
+
+from spark_rapids_pytools.cloud_api.sp_types import CloudPlatform
+from spark_rapids_pytools.common.prop_manager import JSONPropertiesContainer
+from spark_rapids_pytools.common.sys_storage import FSUtil
+from spark_rapids_pytools.common.utilities import Utils
+
+
+# Defines the constants and static configurations
+prepackage_conf = {
+    '_supported_platforms': [csp.value for csp in CloudPlatform if csp != CloudPlatform.NONE],
+    '_configs_suffix': '-configs.json',
+    '_mvn_base_url': 'https://repo1.maven.org/maven2/com/nvidia/rapids-4-spark-tools_2.12',
+    '_folder_name': 'csp-resources'
+}
+
+
+class PrepackageMgr:   # pylint: disable=too-few-public-methods
+    """ Class that handles downloading dependencies to be pre-packaged ahead of runtime.
+
+    The script can be triggered by passing the arguments and the properties of the class
+    For example:
+
+    $> python prepackage_mgr.py --resource_dir=RESOURCE_DIR --tools_jar=TOOLS_JAR run
+
+    :param resource_dir: root folder where the configuration files are located.
+    :param dest_dir: directory in which the resources are downloaded and stored.
+           If missing, the tool creates a subfolder '$resource_dir/csp-resources'.
+           Warning: the 'dest_dir' may be deleted when 'archive_enabled' is set to True.
+    :param tools_jar: path of the spark-rapids-user-tools jar file. Typically, this is the snapshot
+           jar file of the current build.
+    :param archive_enabled: a flag to enable/disable compressing the resources.
```

Comment:
In the documentation, can we specify that by default it is enabled?

Reply:
This is a good point to raise @parthosa . Let me calrify why it is not mentioned in the docstring.

The fact that the default is enabled bu default is emphasized in the default (optional) arguments in the constructor `__init__`.

Explicitly stating the default behavior in the docstring is not recommended as it leads to two problems:

1. It adds redundancy and duplication to the documentation since any python-doc-generator (such as sphinx) automatically generates this info (see the usage command output generated below).
2. It complicates code maintenance. Changing the code (i.e., removing default value) implies that we make changes in two different places: one in the code, and second in the docstring). In most cases, committer/reviewer may not notice that the docstring needs to be updated leading to inconsistent documentation.

Running usage command will generate this information (notice that the output shows the default value and the type of the argument):

```
$> python src/spark_rapids_pytools/resources/dev/prepackage_mgr.py -- --help
# OR
$> python src/spark_rapids_pytools/resources/dev/prepackage_mgr.py --help
```

```
NAME
    prepackage_mgr.py - Class that handles downloading dependencies to be pre-packaged ahead of runtime.

SYNOPSIS
    prepackage_mgr.py --resource_dir=RESOURCE_DIR <flags>

DESCRIPTION
    The script can be triggered by passing the arguments and the properties of the class
    For example:

    $> python prepackage_mgr.py --resource_dir=RESOURCE_DIR --tools_jar=TOOLS_JAR run
    $> python prepackage_mgr.py run --resource_dir=RESOURCE_DIR --tools_jar=TOOLS_JAR

    For more information:

    $> python prepackage_mgr.py --help

ARGUMENTS
    RESOURCE_DIR
        Type: str
        root folder where the configuration files are located.

FLAGS
    -d, --dest_dir=DEST_DIR
        Type: Optional[str]
        Default: None
        directory in which the resources are downloaded and stored. If missing, the tool creates a subfolder '$resource_dir/csp-resources'. Warning: the 'dest_dir' may be deleted when 'archive_enabled' is set to True.
    -t, --tools_jar=TOOLS_JAR
        Type: Optional[str]
        Default: None
        path of the spark-rapids-user-tools jar file. Typically, this is the snapshot jar file of the current build.
    -a, --archive_enabled=ARCHIVE_ENABLED
        Type: bool
        Default: True
        a flag to enable/disable compressing the resources. Note that a wheel-package with compressed prepackaged resources is 30% less size compared to the non-compressed one (~ 600 Vs. 900 MB). When enabled, the prepackaged-resources are stored in '$resource_dir/csp-resources.tgz'. If the 'dest_dir' is provided, then the output is stored as '$dest_dir/../csp-resources.tgz'

```

Reply:
Thanks @amahussein for this clarification. This is a good information.